subscriber_test.go

```diff
@@ -27,6 +29,98 @@ const (
 	updateTimeout = 1 * time.Second
 )
 
+type pubMeta struct {
+	pub legs.Publisher
+	h   host.Host
+}
+
+func TestConcurrentSync(t *testing.T) {
+	err := quick.Check(func(ll llBuilder, publisherCount uint8) bool {
+		return t.Run("Quickcheck", func(t *testing.T) {
+			if publisherCount == 0 {
+				// Empty case
+				return
+			}
+
+			var publishers []pubMeta
+
+			// limit to at most 10 concurrent publishers
+			publisherCount := int(publisherCount)%10 + 1
+
+			for i := 0; i < publisherCount; i++ {
+				ds := dssync.MutexWrap(datastore.NewMapDatastore())
+				pubHost := test.MkTestHost()
+				lsys := test.MkLinkSystem(ds)
+				pub, err := dtsync.NewPublisher(pubHost, ds, lsys, testTopic)
+				if err != nil {
+					t.Fatal(err)
+				}
+				publishers = append(publishers, pubMeta{pub, pubHost})
+
+				head := ll.Build(t, lsys)
+				if head == nil {
+					// We built an empty list. So nothing to test.
+					return
+				}
+
+				err = pub.UpdateRoot(context.Background(), head.(cidlink.Link).Cid)
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			subDS := dssync.MutexWrap(datastore.NewMapDatastore())
+			subLsys := test.MkLinkSystem(subDS)
+			subHost := test.MkTestHost()
+
+			calledTimes := 0
+			sub, err := legs.NewSubscriber(subHost, subDS, subLsys, testTopic, nil, legs.BlockHook(func(i peer.ID, c cid.Cid) {
+				calledTimes++
```

Comment:
Doesn't this need to be synchronized if the same block hook is called concurrently for different publishers?

Reply:
Yes thank you